doc/source/exclude_patterns.md

```diff
@@ -0,0 +1,93 @@
+  - Pattern beginning with double-asterisks will match any leading path segments
+  - Pattern ending with double-asterisks will match any trailing path segments
+  - Pattern with inner double-asterisks will match multiple (or zero) inner path segments 
+
+# Special Characters / Edge Cases 
+1. **Patterns beginning with ' / '**
+  - Match paths on the root directly instead of any descendant paths by making a pattern relative to root. Escape the relative root conversion with an extra forward slash in front so as to match absolute paths. (i.e, ' // '). 
+
+2. **Patterns with a single ' / '**
+  - According to 'git check-ignore' (v2.4.1), a single '/' does not match any file
+
+
+3. **Patterns without beginning ' / '**
+  - A pattern without a beginning slash but contains at least one prepended directory (e.g ``` dir/{pattern} ```) should not match ``` **/dir/{pattern} ``` according to 'git-check-ignore' (v2.4.1.)
+  - A single pattern without a beginning slash will match any descendant path, equivalent to ``` **/{pattern} ```. So, prepend with double-asterisks to make a pattern relative to the root. This also holds for a single pattern with a trailing slash (e.g. ```dir/```). 
+
+4. **Patterns ending with ' / '**
+  - A pattern ending with a slash ('/' will match all descendant paths if it is a directory but not if it is a regular file, equivalent to ``` {pattern}/** ```. So, set the last segment to double asterisks to include all descendants. 
+
+5. **Patterns beginning with ' # '**
+  - Serves as a comment (neither includes or excludes files). Escape the hash with a back-slash to match a literal hash (i.e., '\#'). 
+
+6. **Patterns beginning with ' ! '**
+  - Negates pattern (exclude instead of include). Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e, '\!').
+
+7. **Patterns beginning with ' \\ '**
+  - Removes leading back-slash escape for escape hash or exclamation mark
+
+# Pattern Format Examples
+  - The pattern ``` hello.* ``` matches any file or folder whose name begins with **hello**. If one wants to restrict this only to the directory and not in its subdirectories, one can prepend the pattern with a slash, i.e. ``` /hello.*``` ; the pattern now matches ``` hello.txt, hello.c```  but not ``` a/hello.java ```. 
+  - The pattern ``` foo/ ``` will match a directory **foo** and paths underneath it, but will not match a regular file or a symbolic link **foo** (this is consistent with the way how pathspec works in general in Git)
+  - The pattern ```doc/frotz``` and ```/doc/frotz``` have the same effect as a ```.gitignore file```. In other words, a leading slash is not relevant if there is already a middle slash in the pattern.
+  - The pattern ```foo/*```, matches ```foo/test.json``` (a regular file), ``` foo/bar ``` (a directory), but it does not match ``` foo/bar/hello.c``` (a regular file), as the asterisk in the pattern does not match ``` bar/hello.c```  which has a slash in it.
+
+# Source Code
```

Comment:
I think we don't need the source code section. Perhaps we can include a link, but I'm still wary as we'd have to go by line number which can change. Perhaps just a mention to look in runlib.

Reply:
Agreed, especially since it IS a helper function.